src/provider/provider_tracking.c

```diff
@@ -7,43 +7,84 @@
  *
  */
 
-#include "provider_tracking.h"
+#include <assert.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <umf/memory_pool.h>
+#include <umf/memory_provider.h>
+#include <umf/memory_provider_ops.h>
+
 #include "base_alloc_global.h"
 #include "critnib.h"
 #include "ipc_cache.h"
 #include "ipc_internal.h"
+#include "provider_tracking.h"
 #include "utils_common.h"
 #include "utils_concurrency.h"
 #include "utils_log.h"
 
-#include <umf/memory_pool.h>
-#include <umf/memory_provider.h>
-#include <umf/memory_provider_ops.h>
-
-#include <assert.h>
-#include <errno.h>
-#include <stdio.h>
-#include <stdlib.h>
-#include <string.h>
+#define MAX_LEVELS_OF_MAPS 8
 
 uint64_t IPC_HANDLE_ID = 0;
 
 struct umf_memory_tracker_t {
     umf_ba_pool_t *alloc_info_allocator;
-    critnib *alloc_segments_map;
+    // Multilevel maps are needed to support the case
+    // when one memory pool acts as a memory provider
+    // for another memory pool (nested memory pooling).
+    critnib *alloc_segments_map[MAX_LEVELS_OF_MAPS];
     utils_mutex_t splitMergeMutex;
 };
 
 typedef struct tracker_alloc_info_t {
     umf_memory_pool_handle_t pool;
     size_t size;
+    size_t n_children;
 } tracker_alloc_info_t;
 
 static umf_result_t umfMemoryTrackerAdd(umf_memory_tracker_handle_t hTracker,
                                         umf_memory_pool_handle_t pool,
                                         const void *ptr, size_t size) {
     assert(ptr);
 
+    tracker_alloc_info_t *prev_value = NULL;
+    tracker_alloc_info_t *rvalue = NULL;
+    uintptr_t prev_key = 0;
+    uintptr_t rkey = 0;
+    int prev_level = 0;
+    int level = 0;
+    int found = 0;
+
+    do {
+        found =
+            critnib_find(hTracker->alloc_segments_map[level], (uintptr_t)ptr,
+                         FIND_LE, (void *)&rkey, (void **)&rvalue);
+        if (found && (uintptr_t)ptr < rkey + rvalue->size) {
+            if (level == MAX_LEVELS_OF_MAPS - 1) {
+                LOG_ERR("tracker level is too high, ptr=%p, size=%zu", ptr,
+                        size);
+                return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+            }
+            prev_key = rkey;
+            prev_value = rvalue;
+            prev_level = level;
+            level++;
+        }
+    } while (found && ((uintptr_t)ptr < rkey + rvalue->size) &&
+             rvalue->n_children);
+
+    if (prev_value && ((uintptr_t)ptr + size > prev_key + prev_value->size)) {
```

Comment:
add parentheses to make the code more readable

Reply:
Done